jax/interpreters/pxla.py

```diff
@@ -280,12 +280,27 @@ def axis_env_extend(name, size):
   env = {}
   write(core.unitvar, c.Tuple())
   if const_vals:
+    for val in const_vals:
+      if isinstance(val, xla.DeviceArray):
+        val.copy_to_host_async()
     _map(write, jaxpr.constvars, map(c.Constant, const_vals))
     _map(write, jaxpr.freevars, map(c.ParameterWithShape, freevar_shapes))
   else:
     all_freevars = it.chain(jaxpr.constvars, jaxpr.freevars)
     _map(write, all_freevars, map(c.ParameterWithShape, freevar_shapes))
   _map(write, jaxpr.invars, map(c.ParameterWithShape, arg_shapes))
+
+  # Prefetch any DeviceArray values inside jaxprs to the host.
+  for eqn in jaxpr.eqns:
+    for v in eqn.invars:
+      if type(v) is core.Literal and isinstance(v.val, xla.DeviceArray):
+        v.val.copy_to_host_async()
+    if eqn.bound_subjaxprs:
+      (_, const_bindings, freevar_bindings), = eqn.bound_subjaxprs
```

Comment:
I think we also need to look at the subjaxpr's eqns to find literals there as in the lines above, but the subjaxpr is bound to `_` here and ignored. I think a recursive function `_prefetch_jaxpr_literals` might make sense!

I'm not sure, but I think the lines just below here dealing with `const_bindings` and `freevar_bindings` might never fire; I think both subjaxpr constvars and freevars are instantiated/hoisted, though I'm more certain about the former. (For the former, [this line](https://github.com/google/jax/blob/0a8810d9bc9cfe62307d47f7d71d62dc0a6d8a94/jax/interpreters/partial_eval.py#L115) calls `new_instantiated_const` which creates a hoisted ConstVar rather than a Literal. For the latter, I think env vars are only ever populated by things that originated as Tracers, so [this `full_raise`](https://github.com/google/jax/blob/0a8810d9bc9cfe62307d47f7d71d62dc0a6d8a94/jax/interpreters/partial_eval.py#L116) will also hoist them rather than calling `pure`.) That inference is supported by the fact that in this code we only call c.Constant at the top level when constants are passed in to the jaxpr (around line 286) and for literals that appear in equations.

Reply:
Added a recursive `_prefetch_jaxpr_literals` to xla.py, called it both there and in `pxla`. PTAL.

Reply:
Maybe additional useful context: constants either become Literals (currently only scalars, and my current understanding is only equation invars are ever Literals and not jaxpr invars) or are hoisted all the way to the top level. To handle that hoisting-to-the-top-level in the lowering code (here and in xla.py), while still allowing `replicated_comp` and `jaxpr_computation` to be expressed recursively (and create their own ComputationBuilder instances), there's [a convention](https://github.com/google/jax/blob/0a8810d9bc9cfe62307d47f7d71d62dc0a6d8a94/jax/interpreters/xla.py#L276-L281) in those two functions.

Reply:
Revised version LGTM!